data-gen/src/main/java/io/smartcat/data/loader/SubSetRule.java

```diff
@@ -40,23 +39,24 @@ public boolean isExclusive() {
     }
 
     @Override
-    public Rule<Set<T>> recalculatePrecedance(Rule<Set<T>> exclusiveRule) {
+    public Rule<Collection<T>> recalculatePrecedance(Rule<Collection<T>> exclusiveRule) {
         return null;
     }
 
     @Override
-    public Set<T> getRandomAllowedValue() {
+    public Collection<T> getRandomAllowedValue() {
         return getRandomSubset(values);
     }
 
-    private Set<T> getRandomSubset(Set<T> values) {
+    private Collection<T> getRandomSubset(Collection<T> values) {
         int randomSize = ThreadLocalRandom.current().nextInt(0, values.size());
 
         List<T> list = new ArrayList<>(values);
```

Comment:
This will prove to be a problem with set implementations. We should have separate list and set implementations.